functions/HttpTriggerAPIUsers/__init__.py

```diff
@@ -1,52 +1,136 @@
+
+def get_users(conn):
     with conn.cursor() as cursor:
+        logging.debug(
+            "Using connection cursor to execute query (select all from users)")
         cursor.execute("SELECT * FROM users")
 
         # Get users
-        users = list(cursor.fetchall())
-            
+        logging.debug("Fetching all queried information")
+        users_table = list(cursor.fetchall())
+
         # Clean up to put them in JSON.
-        users = [tuple(user) for user in users]
-            
+        users_data = [tuple(user) for user in users_table]
+
         # Empty data list
-        data = []
+        users = []
+
+        users_columns = [column[0] for column in cursor.description]
+        for user in users_data:
+            users.append(dict(zip(users_columns, user)))
 
-        columns = [column[0] for column in cursor.description]
-        for user in users:
-            data.append(dict(zip(columns, user)))
-            
         # users = dict(zip(columns, rows))
-        return func.HttpResponse(json.dumps(data), status_code=200, mimetype="application/json")
\ No newline at end of file
+        logging.debug(
+            "User data retrieved and processed, returning information from get_users function")
+        return func.HttpResponse(json.dumps(users), status_code=200, mimetype="application/json")
+
+def add_user(conn, user_req_body):
+    # First we want to ensure that the request has all the necessary fields
+    logging.debug("Testing the add new user request body for necessary fields...")
+    try:
+        assert "firstName" in user_req_body, "New user request body did not contain field: 'firstName'"
+        assert "lastName" in user_req_body, "New user request body did not contain field: 'lastName'"
+        assert "email" in user_req_body, "New user request body did not contain field: 'email'"
+    except AssertionError as user_req_body_content_error:
+        logging.error("New user request body did not contain the necessary fields!")
+        return func.HttpResponse(user_req_body_content_error.args[0], status_code=400)
+    
+    logging.debug("New user request body contains all the necessary fields!")
+    with conn.cursor() as cursor:
+        # Unpack user data
+        firstName = user_req_body["firstName"]
+        lastName = user_req_body["lastName"]
+        email = user_req_body["email"]
+
+        # Create the query
+        add_user_query = """
+                         SET NOCOUNT ON;
+                         DECLARE @NEWID TABLE(ID INT);
+
+                         INSERT INTO dbo.users (firstName, lastName, email)
+                         OUTPUT inserted.userId INTO @NEWID(ID)
+                         VALUES('{}', '{}', '{}');
+
+                         SELECT ID FROM @NEWID
+                         """.format(firstName, lastName, email)
+
+
+        logging.debug(
+            "Using connection cursor to execute query (add a new user and get id)")
+        count = cursor.execute(add_user_query)
```

Comment:
[this](https://stackoverflow.com/questions/43491381/pyodbc-the-sql-contains-0-parameter-markers-but-1-parameters-were-supplied-hy0) might be useful in order to avoid injection.

Reply:
I also saw that syntax, but didn't realize it could help with injection attacks! Good catch!